### PR TITLE
Code refactoring and sync from upstream

### PR DESCRIPTION
**Pull request by whitneywhtsang**

### CODE REVIEW

File: third_party/intel/lib/TritonIntelGPUToLLVM/HistogramOpToLLVM.cpp [hunk 1]
```diff
@@ -170,20 +159,22 @@ struct HistogramOpConversion
...
     int numBins = op.getType().getDimSize(0);
-    int numThreadsPerWarp = triton::gpu::TritonGPUDialect::getThreadsPerWarp(
-        op->getParentOfType<ModuleOp>());
-    assert(numThreadsPerWarp < 64 &&
-           "Only supports threads per warp less than 64");
-
+    auto mod = op->getParentOfType<ModuleOp>();
+    int numThreadsPerWarp =
+        triton::gpu::TritonGPUDialect::getThreadsPerWarp(mod);
+    assert(numThreadsPerWarp == 32 ||
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/HistogramOpToLLVM.cpp:165]: We should also allow 16 threads per warp ? Does the codegen consider 16 threads per warp or does this only work for 32/64 threads per warp?

File: third_party/intel/lib/TritonIntelGPUToLLVM/HistogramOpToLLVM.cpp [hunk 2]
```diff
@@ -51,20 +33,22 @@ computeWarpLevelHistogram(Location loc, RankedTensorType srcType,
...
       Value bitSet = and_(value, i32_val(1 << j));
-      Value bit = generateVoteBallot(loc, icmp_ne(bitSet, zero), -1, threadId,
-                                     numThreadPerWarp, rewriter);
+      Value cmp = icmp_ne(bitSet, zero);
+      Value bit =
+          targetInfo.ballot(rewriter, loc, int_ty(numThreadPerWarp), cmp);
       ballotBits.push_back(bit);
     }
-
-    uint64_t fullMaskValue = (1ll << numThreadPerWarp) - 1u;
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/HistogramOpToLLVM.cpp:59]: good point, reverted.

File: third_party/intel/lib/TritonIntelGPUToLLVM/HistogramOpToLLVM.cpp (hunk 1 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/HistogramOpToLLVM.cpp]: fixed, it should work, as the previous assertion was `numThreadsPerWarp < 64`.

File: third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp [hunk 3]
```diff
@@ -14,11 +14,21 @@ using namespace mlir;
 
 namespace mlir::triton::intel {
 
-bool TargetInfo::supportMaximumMinimum() const { return true; }
+bool TargetInfo::supportMaximumMinimum() const { return false; }
 Value TargetInfo::ballot(RewriterBase &rewriter, Location loc, Type type,
                          Value cmp) const {
-  assert("TODO: implement ballot on XPU");
-  return Value();
+  // Emulate vote.ballot.sync behavior using shift, shuffle, and or.
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp:20]: Moved from https://github.com/intel/intel-xpu-backend-for-triton/blob/main/third_party/intel/lib/TritonIntelGPUToLLVM/HistogramOpToLLVM.cpp#L10 with some modifications due to function parameter changes.

File: third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp [hunk 4]
```diff
@@ -14,11 +14,21 @@ using namespace mlir;
 
 namespace mlir::triton::intel {
 
-bool TargetInfo::supportMaximumMinimum() const { return true; }
+bool TargetInfo::supportMaximumMinimum() const { return false; }
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp:17]: Used by `ClampFOpConversion`, need to change to `false` if we want to reuse `ClampFOpConversion` from common files.

**Review by etiotto — APPROVED** [review]

**Pull request merged**